### PR TITLE
GCE pricing model

### PR DESCRIPTION
**Pull request by mwielgus**
ref: #82

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+			} else {
+				priceMapToUse = instancePrices
+			}
+			if basePricePerHour, found := priceMapToUse[machineType]; found {
+				price = basePricePerHour * getHours(startTime, endTime)
+				basePriceFound = true
+			}
+		}
+	}
+	if basePriceFound == false {
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: !basePriceFound

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		price = getBasePrice(node.Status.Capacity, startTime, endTime)
+		if node.Labels != nil && node.Labels[preemptibleLabel] == "true" {
+			price = price * preemptibleDiscount
+		}
+		basePriceFound = true
+	}
+	if !basePriceFound {
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: I can't see how you could ever get into this if.

Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func getAdditionalPrice(resources apiv1.ResourceList, startTime time.Time, endTime time.Time) float64 {
+	if len(resources) == 0 {
+		return 0
+	}
+	hours := getHours(startTime, endTime)
+	price := 0.0
+	gpu := resources[apiv1.ResourceNvidiaGPU]
+	price += float64(gpu.MilliValue()) / 1000.0 * gpuPricePerHour * hours
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Either put this line inside if (based on whether gpu is found in ResourceList), or just declare the variable here.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	hours := getHours(startTime, endTime)
+	price := 0.0
+	cpu := resources[apiv1.ResourceCPU]
+	mem := resources[apiv1.ResourceMemory]
+	price += float64(cpu.MilliValue()) / 1000.0 * cpuPricePerHour * hours
+	price += float64(mem.Value()) / gigabyte * memoryPricePerHourPerGb * hours
+	return price
+}
+
+func getAdditionalPrice(resources apiv1.ResourceList, startTime time.Time, endTime time.Time) float64 {
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Shouldn't we also cover SSD here?

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+				basePriceFound = true
+			}
+		}
+	}
+	if !basePriceFound {
+		price = getBasePrice(node.Status.Capacity, startTime, endTime)
+		if node.Labels != nil && node.Labels[preemptibleLabel] == "true" {
+			price = price * preemptibleDiscount
+		}
+		basePriceFound = true
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: This line can be removed

File: cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	node1.Labels = labels1
+	price1, err := model.NodePrice(node1, now, now.Add(time.Hour))
+	assert.NoError(t, err)
+
+	// preemptable
+	node2 := BuildTestNode("sillyname2", 8000, 30*1024*1024*1024)
+	node2.Labels = labels2
+	price2, err := model.NodePrice(node2, now, now.Add(time.Hour))
+	assert.NoError(t, err)
+	assert.True(t, price1 > 3*price2)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go]: I would rather base this assert on preemptibleDiscount const or at the very least make 3 a const. I don't like an idea of unittests failing after a change in pricing model (especially if it was taken from config file as per your TODO)

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: if gpu is not found then gpu value is 0.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: I'm still thinking how to handle SSD the best. Adding todo.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go (hunk 6 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go]: This is only a sanity check. I didn't want to base any of the checks on real numbers.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+	assert.True(t, price4 > price5)
+	assert.True(t, price4 < 1.5*price5)
+	assert.True(t, price4 > 2*price1)
+
+	// small custom node
+	node6 := BuildTestNode("sillyname6", 1000, 3750*1024*1024)
+	price6, err := model.NodePrice(node6, now, now.Add(time.Hour))
+	assert.NoError(t, err)
+	assert.True(t, math.Abs(price3-8*price6) < 0.1)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go]: This check looks really magical.

Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model_test.go]: 6 is 8 times smaller than 3. We are doing float point calculation so no equality check can be made here.

File: cluster-autoscaler/cloudprovider/gce/gce_price_model.go (hunk 3 above)
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_price_model.go]: Yeah, I get that. Just meant it looks somewhat strange to do:
```
price := 0.0
price += something
```
as opposed to 
```
price := something
```
unless we have different execution paths. I guess it doesn't really matter though.

**Pull request merged**